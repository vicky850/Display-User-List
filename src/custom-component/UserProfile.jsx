/**
 * Summary: User Profile
 * Description: User profile (stateless component)
 * @author Vikash Kumar.
 * @date  28.07.2018
 */

import React from "react";
import { actionTypes } from '../constants/action-type.constants';
require('./css/user-profile.css');

export const UserProfile = (props) => {
    return (
      props.userlist.length ? props.userlist.map((user, index) => {
        return <div key={"box" + index} className="box">
          <div className="profile-image"><img src={user.avatar} alt="" /></div>
          <h2>{user.first_name} {user.last_name}</h2>
          <span className="delete" data-action={actionTypes.DELETE_DATA} onClick={(e)=>props.optionHandler(e, index)}>Delete</span>
        </div>
      }) : "Please refresh the page to load the data !"
    )
  }