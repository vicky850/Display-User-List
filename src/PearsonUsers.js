/**
 * Summary: PersonUsers component which render the UserProfile list
 * Description: PersonUsers component which render the UserProfile list
 * @author Vikash Kumar.
 * @date  28.07.2018
 */
import React, { Component } from "react";
import { UserProfile } from './custom-component/UserProfile'
import { actionTypes } from './constants/action-type.constants';
import { alertConstants } from './constants/alert.constants';
import { customErrorConst } from './constants/error.constants';
import { userActions } from './actions/user.actions';
import _ from 'lodash';
require('./style.css');

//Pearson User Class
export class PearsonUsers extends Component {
  constructor(props) {
    super(props);
    this.optionHandler = this.optionHandler.bind(this)

    this.state = {
      users: [
        {
          id: 4,
          first_name: "Eve",
          last_name: "Holt",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        },
        {
          id: 5,
          first_name: "Charles",
          last_name: "Morris",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
        },
        {
          id: 6,
          first_name: "Tracey",
          last_name: "Ramos",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
        }
      ],
      alert: { type: null, message: null },
      page: 1,
      per_page: 10
    };
  }

  optionHandler(ev, id, index) {
    if (ev.target.dataset.action === actionTypes.DELETE_DATA) {
      var newUserList = [...this.state.users]; // make a separate copy of the array
      newUserList.splice(index, 1);
      this.setState({ users: newUserList, page: 0 });
    }
  }

  displayUsers(e) {
    let newUserList = [];
    let pageCount = this.state.page;
    //get user list
    userActions.getUsers(pageCount, this.state.per_page).then(
      (response) => {
        if (response) {
          if (response.type && (response.type === alertConstants.SUCCESS)) {
            response = response.message;
            newUserList = response.data.data;
            let error_alert = { type: null, message: null };
            this.setState({ users: [...this.state.users, ...newUserList], page: pageCount + 1, alert: error_alert });
          } else if (response.type === alertConstants.ERROR) {
            let error_alert = { type: alertConstants.ERROR, message: customErrorConst.ERROR_DEFAULT };
            this.setState({ alert: error_alert });
          }
        } else {
          let error_alert = { type: alertConstants.ERROR, message: customErrorConst.ERROR_DEFAULT };
          this.setState({ alert: error_alert });
        }
      }
    ).catch(this.errorHandle);
  }

  removeDuplicate(e) {
    let uniqueUsers = _.mapKeys(this.state.users, 'id');
    let uniqueUsersArray = _.map(uniqueUsers);
    this.setState({ users: uniqueUsersArray });
  }


  render() {
    return (
      <div className="pearon-users">
        <h1>Pearson User Management</h1>
        {
          this.state.alert.message &&
          <div id="alert-messgae" className={`alert alert-${this.state.alert.type}`}>{this.state.alert.message}</div>
        }
        <div className="bx-wrapper">
          <UserProfile optionHandler={this.optionHandler} userlist={this.state.users} />
        </div>
        <div>
          <button type="button" className="button" id="fetchUsers" onClick={(e) => this.displayUsers(e)}>
            <span className="blue" aria-hidden="true"></span> Get More Users
          </button>
          <button type="button" className="button" id="fetchUsers" onClick={(e) => this.removeDuplicate(e)}>
            <span className="blue" aria-hidden="true"></span> Remove Duplicate
          </button>
        </div>

      </div>
    );
  }
}
