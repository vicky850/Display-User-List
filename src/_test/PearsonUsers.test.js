import React from 'react';
import { PearsonUsers } from "../PearsonUsers";
import { shallow } from 'enzyme';
import { UserProfile } from '../custom-component/UserProfile';

//test case for rendring user list and elements
describe('PearsonUsers Test Suite', function () {
  let component;
  beforeEach(() => {
    component = shallow(<PearsonUsers />);
  })
  it('should render without throwing any error', function () {
    expect(component.length).toEqual(1)
  })

  it("renders a h1", () => {
    const h1 = component.find("h1");
    expect(h1.text()).toEqual("Pearson User Management");
  });

  it("should render initial layout", () => {
    expect(component.getElements()).toMatchSnapshot();
  });



  //data set for testing
  let _statePearsonUsers = [
    {
      id: 4,
      first_name: "Eve",
      last_name: "Holt",
      avatar:
        "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
    },
    {
      id: 5,
      first_name: "Charles",
      last_name: "Morris",
      avatar:
        "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
    },
    {
      id: 5,
      first_name: "Charles",
      last_name: "Morris",
      avatar:
        "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
    }
  ]

  var objPearsonUsers = new PearsonUsers();
  
  //mocking of setState function
  PearsonUsers.prototype.setState = obj => {
    if (typeof obj === 'function') {
      var _obj = obj(stateData);
      if (Object.keys(_obj).length > 0) {
        var testKey = Object.keys(_obj);
        for (var i = 0; i < testKey.length; i++) {
          objPearsonUsers.state[testKey[i]] = _obj[testKey[i]];
        }
      }
      return _obj;
    }
    if (Object.keys(obj).length > 0) {
      var testKey = Object.keys(obj);
      for (var i = 0; i < testKey.length; i++) {
        objPearsonUsers.state[testKey[i]] = obj[testKey[i]];
      }
    }
    return obj;
  };

  it("it should remove duplicate entries", () => {
    component.setState({ users: _statePearsonUsers });
    objPearsonUsers.removeDuplicate();
    expect(objPearsonUsers.state.users.length).toBe(2);
  });


  it("should delete individual user", () => {
    let event={target:{dataset :{action:"delete"}}};
    objPearsonUsers.optionHandler(event, 1);
    expect(objPearsonUsers.state.users[0]).toEqual(_statePearsonUsers[0]);
  });
})
