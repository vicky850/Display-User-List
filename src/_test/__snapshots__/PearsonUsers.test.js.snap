// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PearsonUsers Test Suite should render initial layout 1`] = `
Array [
  <div
    className="pearon-users"
>
    <h1>
        Pearson User Management
    </h1>
    <div
        className="bx-wrapper"
    >
        <Unknown
            optionHandler={[Function]}
            userlist={
                Array [
                    Object {
                      "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                      "first_name": "Eve",
                      "id": 4,
                      "last_name": "Holt",
                    },
                    Object {
                      "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                      "first_name": "Charles",
                      "id": 5,
                      "last_name": "Morris",
                    },
                    Object {
                      "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                      "first_name": "Tracey",
                      "id": 6,
                      "last_name": "Ramos",
                    },
                  ]
            }
        />
    </div>
    <div>
        <button
            className="button"
            id="fetchUsers"
            onClick={[Function]}
            type="button"
        >
            <span
                aria-hidden="true"
                className="blue"
            />
             Get More Users
        </button>
        <button
            className="button"
            id="fetchUsers"
            onClick={[Function]}
            type="button"
        >
            <span
                aria-hidden="true"
                className="blue"
            />
             Remove Duplicate
        </button>
    </div>
</div>,
]
`;
